### PR TITLE
ClusterWriter/ClusterListener with connection pooling.

### PR DESCRIPTION
**Pull request by corylanou**
This introduces the ClusterWriter and the ClusterListener.

### CODE REVIEW

File: tcp/client.go [hunk 1]
```diff
@@ -5,32 +5,83 @@ import (
...
+			return nil, err
+		}
+		c.pool.setPool(addr, p)
+	}
+	return c.pool.conn(addr)
 }
 
-func (c *Client) Dial(addr string) error {
-	conn, err := net.Dial("tcp", addr)
+func (c *Client) WriteShard(addr string, shardID uint64, points []tsdb.Point) error {
```
Comment (pauldix) [inline tcp/client.go]: I would't expect `WriteShard` to take a connection address. I would expect a `Client` to be tied to a specific address already and `WriteShard` to simply take the `shardID` and the `points`.

File: tcp/conn.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package tcp
+
+import (
+	"net"
+
+	"github.com/davecgh/go-spew/spew"
```
Comment (pauldix) [inline tcp/conn.go]: leftover I assume?

**Comment by jwilder** [comment]
LGTM :+1:

**Comment by pauldix** [comment]
The organization of where pools are is a little odd to me. When I think of a connection pool, I generally think of that as a pool to a single remote server. i.e. connections to the same host/port.

I would expect the `cluster.Writer` to keep the map of data node to the `pool` objects.

**Comment by pauldix** [comment]
Also, it's better to key those pool objects off the `nodeID` rather than the `*NodeInfo`. When the write comes in it always comes with the `nodeID`.

You check if the pool is in the map, if so great, use it. If not, then get the `*NodeInfo` from `Metastore` so that you can get the connection string to dial.

**Pull request merged**